### ISSUE TITLE
Exec fails with no such file or directory

### ISSUE THREAD

**Issue by sashayakovtseva** [apptainer/singularity#2386]
## Version of Singularity:

3.0.1-112.g5655f87.dirty

### Expected behavior

Command is executed inside a container successfully.

### Actual behavior

Command execution results in `exec /.singularity.d/actions/exec failed: no such file or directory`.

### Steps to reproduce behavior

1. sudo singularity -d build echo.sif docker://hashicorp/http-echo
2. singularity exec echo.sif ls

Output: 
```
WARNING: passwd file doesn't exist in container, not updating
WARNING: group file doesn't exist in container, not updating
FATAL:   exec /.singularity.d/actions/exec failed: no such file or directory
```

**Comment by sashayakovtseva** [comment]
The interesting part is that building go echo server from source (https://github.com/hashicorp/http-echo) and then building SIF image natively works.

**Comment by cclerget** [comment]
@sashayakovtseva `no such file or directory` here means there is no `/bin/sh` inside image, same things happen with a very minimal container like :

```
singularity shell docker://hello-world
```

Since those shell scripts rely on `/bin/sh` presence in image, some images are actually unusable with Singularity execution logic

**Comment by jscook2345** [comment]
@cclerget 

Is there a way to improve this error?

    FATAL:   exec /.singularity.d/actions/shell failed: no such file or directory

seems worse than these

    -bash: ./tmp: /usr/bin/bob: bad interpreter: No such file or directory

    sh: ./tmp: /usr/bin/bob: bad interpreter: No such file or directory

**Comment by cclerget** [comment]
@jscook2345 Yeah this is possible, that will require some code after error to inspect file in order to detect file type and handle error correctly, because `no such file or directory` is also returned for a missing shared library

**Comment by phpHavok** [comment]
@jscook2345 @cclerget I'm wondering how complicated this needs to be. I've added a very simple PR that seems to solve this on the surface, but I'm sure I'm missing details.
```
[jacob@gideon temp]$ singularity exec echo.sif ls
WARNING: passwd file doesn't exist in container, not updating
WARNING: group file doesn't exist in container, not updating
FATAL:   /bin/sh doesn't exist in container: no such file or directory
```

**Comment by mem** [comment]
The reason why the error says `no such file or directory` is because the container image does not have `/bin/sh` (presumably because nothing in the image needs it).

The reason why singularity needs `/bin/sh` is because we are not running the command (`ls` in this case) directly, but instead we are using a script to launch it. That's why the error talks about `/.singularity.d/actions/exec` and `/.singularity.d/actions/shell` (those are the wrappers). Containers where `/bin/sh` does not exist will all have this problem...

Maybe we need to check for the existence of `/bin/sh` in the image before we try to execute the command in order to produce a better error message? Or we need to turn those wrappers into static binaries so we don't have external dependencies?

**Comment by cclerget** [comment]
@phpHavok This issue is not as simple as it appears. As @mem mentioned, check may happen early in `checkExec` function by example, but a missing `/bin/sh` in container is not the only reason to get a `no such file or directory` error, a missing shared library, a binary for another architecture (eg: `singularity shell docker://ppc64le/ubuntu`) will generate this error too even with a `/bin/sh` inside image, so to not mislead user about error we need to make distinction and add checks to report the right cause :

- if `/bin/sh` is missing, display `no /bin/sh` error
- if `/bin/sh` is present, check architecture match :
  - if not, display something like `image targets xxxx architecture`
  - if match, take the lazy approach by assuming that a shared library is missing and just display `a shared library is missing in image` error without telling user the exact library, otherwise it involves to get/search for all library dependencies by opening each binary files, `/bin/sh` and all encountered shared libraries

**Comment by phpHavok** [comment]
@cclerget I think I have a working version with those suggestions implemented. PR has been updated.

**Comment by cclerget** [comment]
Fixed by #2918
